apply unicode normalization in help mode

epithet opened the pull request:
This adapts the help mode to recent changes from #40948 and #25157.

Before:
```
help?> −
"−" can be typed by \minus<tab>

search:

  No documentation found.

  Binding − does not exist.
```

Now:
```
help?> −
"−" can be typed by \minus<tab>

search: - ->

  -(x)

  Unary minus operator.
  ...
```

Same for `help?> \cdotp<tab>` when `\cdot` has been bound.

Note that `help?> 1e−2` already worked.

epithet committed:
apply unicode normalization in help mode

epithet committed:
add tests for unicode normalization in help mode

vtjnash commented on stdlib/REPL/src/docview.jl:
```diff
@@ -452,13 +452,21 @@ function repl_latex(io::IO, s::String)
 
 function repl(io::IO, s::Symbol; brief::Bool=true)
+    str_orig = string(s)
+    s = normalize_symbol(s)
```
Who is constructing symbols explicitly, without having parsed them?

epithet replied:
That's quite early in `_helpmode(io::IO, line::AbstractString)`. The whole input line is made into a symbol (apart from stripped white space at the ends):
https://github.com/JuliaLang/julia/blob/e66091887b2882e2df7324fd7710a46aa75e0289/stdlib/REPL/src/docview.jl#L48
In the case of `\minus`, `\cdot`, and `\cdotp`, the if branch directly afterwards is executed because `Base.isoperator(assym) == true`, and so `@repl` is called with `assym`, which in turn calls `repl(io::IO, s::Symbol; brief::Bool=true)`.

epithet replied:
The information which exact characters were entered is needed at that point for `repl_latex($io, $str_orig)`, which shows
```
"−" can be typed by \minus<tab>
```

So, constructing the symbol without parsing seems to be a sort of trick to carry two kinds of information (the exact input and the fact that it will be parsed as a keyword/operator/...) in one macro/function argument on which we can dispatch.

epithet replied:
I just dug a bit deeper and found #19464. µ (U+00B5 micro) and μ (U+03BC greek small letter mu) both show
```
"μ" can be typed by \mu<tab>
```
which is technically only true for the latter. This is because the code goes through the `else` branch, so the symbol is not `assym` but `x` which comes from `Meta.parse`. But the binding is shown correctly, which is much more important.

In it's current state, this PR is really only a quick fix for specific cases which behave clearly wrong in my opinion. For anything further, the desired behavior should be decided on first.

epithet replied:
And I found out why the symbols are constructed explicitly without parsing. That was introduced by #35154 because of this:
```julia
julia> x = Meta.parse("+=", raise = false, depwarn = false)
:($(Expr(:error, "invalid identifier name \"+=\"")))

julia> x = Meta.parse(".=", raise = false, depwarn = false)
:($(Expr(:error, "invalid identifier name \".=\"")))
```

epithet replied:
Something like this could be better for `+=` and `.=`, what do you say?
```julia
x = Meta.parse(line, raise = false, depwarn = false)
if Meta.isexpr(x, :error)
    asinfix = Meta.parse("x "*line*" x", raise = false, depwarn = false)
    if asinfix isa Expr && length(asinfix.args) == 2 && asinfix.args[1] == asinfix.args[2] == :x    
        x = asinfix.head
    end   
end   
```
I'm not entirely sure if that breaks other things.

epithet committed:
clean up help mode, add tests, and fix \xor=

epithet commented:
I cleaned up a bit according to my findings. It seems to work well. We get LaTeX tab-completion help now for apropos:
```
help?> "λ"
""λ"" can be typed by "\lambda<tab>"

Base.codeunits
...
```

And floating-point literals are also fixed (I was mistaken in the beginning):
```
help?> 1e−2
"1e−2" can be typed by 1e\minus<tab>2

  Float64(x [, mode::RoundingMode])
...
```

µ (U+00B5 micro) and μ (U+03BC greek small letter mu) are technically more correct, now. I'm not sure what should be displayed in this case, we could also use the symbol if that could be parsed:
```
help?> µμ
"µμ" can be typed by µ\mu<tab>
...
```

epithet commented:
Now, the following cases are handled: `−`, `−=`, `.−`, `.−=`, `1e−2`.
And I also fixed `⊻=`, which was already broken in 1.6.1.

vtjnash commented on stdlib/REPL/src/docview.jl:
```diff
@@ -45,26 +36,52 @@ function _helpmode(io::IO, line::AbstractString)
+    if Meta.isexpr(x, :error)
+        # handle operators like +=
+        asinfix = Meta.parse("x $line x", raise = false, depwarn = false)
+        if asinfix isa Expr && length(asinfix.args) == 2 && asinfix.args[1] == asinfix.args[2] == :x
```
```suggestion
        if asinfix isa Expr && length(asinfix.args) == 2 && asinfix.args[1] === asinfix.args[2] === :x
```

vtjnash commented on stdlib/REPL/src/docview.jl:
```diff
@@ -45,26 +36,52 @@ function _helpmode(io::IO, line::AbstractString)
+            x = asinfix.head
         end
+    end
+    if Meta.isexpr(x, :.) && length(x.args) == 1 && x.args[1] isa Symbol
```
```suggestion
    if Meta.isexpr(x, :., 1) && x.args[1] isa Symbol
```

vtjnash commented on base/docs/Docs.jl:
```diff
@@ -312,7 +312,7 @@ astname(@nospecialize(other), ismacro::Bool) = other
 macroname(x::Expr)   = Expr(x.head, x.args[1], macroname(x.args[end].value))
 
-isfield(@nospecialize x) = isexpr(x, :.) &&
+isfield(@nospecialize x) = isexpr(x, :.) && length(x.args) == 2 &&
```
```suggestion
isfield(@nospecialize x) = isexpr(x, :., 2) &&
```

vtjnash commented on stdlib/REPL/src/docview.jl:
```diff
@@ -25,17 +25,8 @@ helpmode(line::AbstractString) = helpmode(stdout, line)
 
-function _helpmode(io::IO, line::AbstractString)
+function _helpmode_parse(line::AbstractString)
     line = strip(line)
```
```suggestion
```

vtjnash commented on stdlib/REPL/src/docview.jl:
```diff
@@ -45,26 +36,52 @@ function _helpmode(io::IO, line::AbstractString)
+    end
+    if Meta.isexpr(x, :.) && length(x.args) == 1 && x.args[1] isa Symbol
+        # handle broadcast operators
+        x = Symbol(string(x.head) * string(x.args[1]))
```
```suggestion
        x = Symbol(".", x.args[1]::Symbol)
```

vtjnash reviewed (approved):
Some minor suggestions to improve clarity in some places, but otherwise should be great to merge now. Thanks!

vtjnash commented on stdlib/REPL/src/docview.jl:
```diff
@@ -221,7 +238,7 @@ function lookup_doc(ex)
         isdotted = startswith(str, ".")
         if endswith(str, "=") && Base.operator_precedence(ex) == Base.prec_assignment && ex !== :(:=)
-            op = str[1:end-1]
+            op = str[1:prevind(str, end, 1)]
```
already fixed, so this is now causing a merge conflict
```suggestion
            op = str[1:end-1]
```

vtjnash committed:
Apply suggestions from code review

vtjnash commented:
Decided simply to apply my comments, since it was all minor stuff, so that this can be merged as soon as CI finishes (if I didn't type something wrong)

DilumAluthge commented:
CI looks... very sad.

DilumAluthge commented:
I'm removing the `merge me` label, because the CI failures are related to the contents of the PR.
